include/alpaka/pltf/PltfGenericSycl.hpp

```diff
@@ -18,89 +18,150 @@
 
 #    include <CL/sycl.hpp>
 
-#    include <iostream>
+#    include <cstddef>
+#    include <exception>
+#    include <functional>
+#    if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
+#        include <iostream>
+#    endif
+#    include <mutex>
+#    include <optional>
+#    include <shared_mutex>
 #    include <sstream>
 #    include <stdexcept>
 #    include <vector>
 
 namespace alpaka::experimental
 {
     //! The SYCL device manager.
-    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl>
+    template<typename TSelector>
+    class PltfGenericSycl : public concepts::Implements<ConceptPltf, PltfGenericSycl<TSelector>>
     {
     public:
         PltfGenericSycl() = delete;
+
+        [[nodiscard]] static auto syclPlatform() -> sycl::platform&
+        {
+            auto lock = std::unique_lock<std::mutex>{platform_mutex};
+
+            if(!platform_opt.has_value())
+                platform_opt = std::make_optional<sycl::platform>(TSelector{});
+
+            return platform_opt.value();
+        }
+
+        [[nodiscard]] static auto syclDevices() -> std::vector<sycl::device>&
+        {
+            auto lock = std::unique_lock<std::mutex>{devices_mutex};
+
+            if(!devices_opt.has_value())
+                devices_opt = std::make_optional<std::vector<sycl::device>>(syclPlatform().get_devices());
```

Comment:
```suggestion
                devices_opt = syclPlatform().get_devices();
```
should work as well ?